### ISSUE TITLE
How to get penultimate hidden layer output after fitting model?

### ISSUE THREAD

**Issue by MichaelMMeskhi** [skorch-dev/skorch#705]
Hi, 

I am trying to train a model. After it is done training and say I am happy with the results. I would like to pass another input and extract the output of the last hidden layer and not the output layer of the trained model. How can I achieve this?

**Comment by BenjaminBossan** [comment]
There are a few ways you could do it. E.g. you could use an attribute on the module to signal whether you want the output from last or from the second to last layer. That could look like this:

```python
class MyModule(nn.Module):
    def __init__(self, ..., apply_output_layer=True):
        super().__init__()
        ...
        self.apply_output_layer = apply_output_layer

    def forward(self, X):
        out = ...
        if self.apply_output_layer is True:
            return self.output_layer(out)
        return out

net = NeuralNet(MyModule, ...)
net.fit(X, y)
# switch predict mode
net.set_params(module__apply_output_layer=False)
y_out = net.predict(out)
```

Instead of using a custom attribute like `apply_output_layer`, you could use the `training` attribute that is already present from PyTorch. The advantage of that is that you don't need to introduce a new attribute. The disadvantage is that for the skorch internal validation, `training` is also set to `False`, so internal validation may break.

**Comment by MichaelMMeskhi** [comment]
@BenjaminBossan Thanks for getting back to me so quickly! Yes, I was playing around with the built in `training` parameter and realized it breaks the internal validation cycle. I will go ahead with your suggestion!

**Comment by MichaelMMeskhi** [comment]
Hi, 

```
class AbstractNet(nn.Module):
    def __init__(self, input_dim, hidden_dim, latent_dim, 
                 output_dim, non_lin, apply_output_layer=True):
        super(AbstractNet, self).__init__()

        # Options
        self.non_lin = non_lin
        self.apply_output_layer = apply_output_layer
        
        # Architecture
        self.input = nn.Linear(input_dim, hidden_dim)
        self.hidden_1 = nn.Linear(hidden_dim, hidden_dim)
        self.hidden_2 = nn.Linear(hidden_dim, hidden_dim)
        self.hidden_3 = nn.Linear(hidden_dim, hidden_dim)
        self.hidden_4 = nn.Linear(hidden_dim, latent_dim)
        self.hidden_5 = nn.Linear(latent_dim, latent_dim)
        self.output = nn.Linear(latent_dim, output_dim)

    def forward(self, X, **kwargs):
        X = self.non_lin(self.input(X))
        X = self.non_lin(self.hidden_1(X))
        X = self.non_lin(self.hidden_2(X))
        X = self.non_lin(self.hidden_3(X))
        X = self.non_lin(self.hidden_4(X))
        X = self.non_lin(self.hidden_5(X))
    
    
        if self.apply_output_layer is True:
            return torch.sigmoid(self.output(X))
        
        return X
```

This is my sample network using the `apply_outer_layer` flag. I do the following:

1. Fit network on training
2. Score model on training data
2. Disable output layer
3. Output predictions of training data from last hidden layer 
4. Enable `apply_output_layer`
5. Score model on validation data

There is an issue I noticed. When scoring network on training then doing steps 3-5, validation score is bad. But when I score model right after step 2 (after training score), validation score is good. Does doing `net.set_params(module__apply_output_layer=False)` somehow change the weights or something?

**Comment by BenjaminBossan** [comment]
Oh yes, this is what happens, my bad. Using `set_params` this way re-initializes the module (there should be a print output indicating this). This is because skorch doesn't know if the module needs to be re-initialized after the change. The correct way to solve this is using the following instead of `set_params`:

`net.module_.apply_output_layer = False`

@ottonemo Could this be solved with virtual params somehow?

**Comment by MichaelMMeskhi** [comment]
Awesome! That's what I noticed that the parameters were reset and I had no idea how to prevent that. Thank you.

**Comment by MichaelMMeskhi** [comment]
Sometimes I get an error, `module_` not found?

**Comment by BenjaminBossan** [comment]
`module_` should always exist after fitting the model, but not before.
